### PR TITLE
kulikov lab2 var 3

### PR DESCRIPTION
**Pull request by AveZorgen**
Реализован плагин
Добавлены CMakeLists для подтягивания написанных плагинов в тесты
вот так вот

### CODE REVIEW

File: llvm/labs/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FILE(GLOB children RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/*)
+FOREACH(child ${children})
+  IF(IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/${child})
+    add_subdirectory(${child})
+  ENDIF()
+ENDFOREACH()
+
+set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
```
Comment (m-ly4) [inline llvm/labs/CMakeLists.txt]:
```suggestion
set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)

```

File: llvm/labs/lab1/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+FILE(GLOB children RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/*)
+FOREACH(child ${children})
+  IF(IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/${child})
+    add_subdirectory(${child})
+  ENDIF()
+ENDFOREACH()
+
+set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
```
Comment (m-ly4) [inline llvm/labs/lab1/CMakeLists.txt]:
```suggestion
set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)

```

File: llvm/test/lab1/kulikov_artem/3for.ll [hunk 3]
```diff
@@ -0,0 +1,148 @@
+; RUN: opt -passes=kulikov-wrap-plugin -S %s | FileCheck %s
+
+@j = dso_local global i32 2, align 4
+
```
Comment (m-ly4) [inline llvm/test/lab1/kulikov_artem/3for.ll]: Would you mind to provide in comments in the test a c++ code you used for generation of input LLVM IR?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/labs/lab1/kulikov_artem/Cool.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    for (auto &L : LI) {
+      auto *Preheader = L->getLoopPreheader();
+      if (!Preheader) {
+        continue;
+      }
+      IRBuilder<> Builder(Preheader->getTerminator());
+      auto *LoopStartFunc = F.getParent()->getFunction("loop_start");
+
+      if (!LoopStartFunc) {
+        LoopStartFunc = Function::Create(
```
Comment (m-ly4) [inline llvm/labs/lab1/kulikov_artem/Cool.cpp]: You can use just `Module::getOrInsertFunction` method.

File: llvm/test/lab1/kulikov_artem/3for.ll [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+;     a();
+;     for (; ; j++) {
+;         if (j > 5)
+;             break;
+;     }
+;     b(c);
+;     int q = c + 42;
+;     return q;
+; }
+
```
Comment (m-ly4) [inline llvm/test/lab1/kulikov_artem/3for.ll]: Could you please also add a test for `while` loop and `if` conditional?

File: llvm/test/lab2/kulikov_artem/3for.ll [hunk 6]
```diff
@@ -0,0 +1,150 @@
+; RUN: opt -passes=kulikov-wrap-plugin -S %s | FileCheck %s
```
Comment (m-ly4) [inline llvm/test/lab2/kulikov_artem/3for.ll]: Could you please load this plugin dynamically via `-load-pass-plugin` option of opt?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab1/kulikov_artem/3for.ll (hunk 5 above)
Comment (AveZorgen) [reply llvm/test/lab1/kulikov_artem/3for.ll]: If inside a while loop like in v function?

File: llvm/test/lab2/kulikov_artem/3for.ll (hunk 6 above)
Comment (AveZorgen) [reply llvm/test/lab2/kulikov_artem/3for.ll]: This plugin is built in static library. We can't load it

Comment (m-ly4) [reply llvm/test/lab2/kulikov_artem/3for.ll]: Making it dynamic is part of the comment.

File: llvm/test/lab1/kulikov_artem/3for.ll (hunk 5 above)
Comment (m-ly4) [reply llvm/test/lab1/kulikov_artem/3for.ll]: I meant a function just with if-else statement to check that there is no loop guarding calls added.

Comment (AveZorgen) [reply llvm/test/lab1/kulikov_artem/3for.ll]: ok

**Review by m-ly4 — APPROVED** [review]
Thank you for applies.

**Pull request merged**